designer/server/src/main/scala/pl/touk/nussknacker/ui/process/migrate/RemoteEnvironment.scala

```diff
@@ -195,25 +216,73 @@ trait StandardRemoteEnvironment extends FailFastCirceSupport with RemoteEnvironm
       processName: ProcessName,
       isFragment: Boolean
   )(implicit ec: ExecutionContext, loggedUser: LoggedUser): Future[Either[NuDesignerError, Unit]] = {
-    val migrateScenarioRequest: MigrateScenarioRequest =
-      MigrateScenarioRequest(
-        environmentId,
-        processingMode,
-        engineSetupName,
-        processCategory,
-        scenarioGraph,
-        processName,
-        isFragment
+
+    val result: EitherT[Future, NuDesignerError, Unit] = for {
+      remoteScenarioDescriptionVersion <- fetchRemoteMigrationScenarioDescriptionVersion
+      localScenarioDescriptionVersion = migrationApiAdapterService.getCurrentApiVersion
+      migrateScenarioRequest: MigrateScenarioData =
+        MigrateScenarioDataV2(
+          environmentId,
+          loggedUser.username,
+          processingMode,
+          engineSetupName,
+          processCategory,
+          scenarioGraph,
+          processName,
+          isFragment
+        )
+      versionsDifference = localScenarioDescriptionVersion - remoteScenarioDescriptionVersion
+      transformedMigrateScenarioRequestE =
+        if (versionsDifference > 0)
+          migrationApiAdapterService.adaptDown(migrateScenarioRequest, versionsDifference)
+        else
+          Right(migrateScenarioRequest)
+      _ <- handleTransformedMigrateScenarioRequest(transformedMigrateScenarioRequestE)
+    } yield ()
+
+    result.value
+  }
+
+  private def handleTransformedMigrateScenarioRequest(
+      transformedMigrateScenarioRequestE: Either[ApiAdapterServiceError, MigrateScenarioData]
+  )(implicit ec: ExecutionContext): EitherT[Future, NuDesignerError, Unit] = {
+    transformedMigrateScenarioRequestE match {
+      case Left(apiAdapterServiceError) =>
+        EitherT.leftT(MigrationApiAdapterError(apiAdapterServiceError))
+      case Right(transformedMigrateScenarioRequest) =>
+        val dto = MigrateScenarioData.fromDomain(transformedMigrateScenarioRequest)
+        EitherT(
+          invokeForSuccess(
+            HttpMethods.POST,
+            List("migrate"),
+            Query.Empty,
+            HttpEntity(dto.asJson.noSpaces),
+            List()
+          )
+        )
+    }
+  }
+
+  private def fetchRemoteMigrationScenarioDescriptionVersion(implicit ec: ExecutionContext) = {
+    EitherT(
+      fetchRemoteMigrationScenarioDescriptionVersionAux.map[Either[NuDesignerError, Int]](apiVersion =>
+        Right(apiVersion.value)
       )
-    invokeForSuccess(
-      HttpMethods.POST,
-      List("migrate"),
-      Query.Empty,
-      HttpEntity(migrateScenarioRequest.asJson.noSpaces),
-      List.empty
     )
   }
 
+  private def fetchRemoteMigrationScenarioDescriptionVersionAux(implicit ec: ExecutionContext): Future[ApiVersion] = {
+    invoke[ApiVersion](
```

Comment:
let's leave a todo, that we could use a client that is generated from Tapir endpoint

Reply:
please take a look at my improvement suggestions: 
https://github.com/TouK/nussknacker/pull/5975